[2.0] Allow dynamic configuration of data sources

martijnwalraven opened the pull request:
This PR adds callbacks to `RESTDataSource` that allow for asynchronous and dynamic configuration of data sources. It addresses #1181 and #1238.

It allows a data source to specify `baseURL`, `defaults` and `defaultParams` either as simple properties, properties that return a `Promise`, or as a function that takes `RequestOptions` and returns either a value or a `Promise`.

So you could implement the example from https://github.com/apollographql/apollo-server/issues/1181#issuecomment-398871456 as follows:

```typescript
class FooAPI extends RESTDataSource {
  baseURL = (options: RequestOptions) => {
    return new Promise<string>((resolve, reject) => {
      dns.resolveSrv(options.path.split("/")[1] + ".service.consul", (err, addresses) => {
        err ? reject(err) : resolve("http://" + addresses[0].name + ":" + addresses[0].port);
      });
    });
  }

  getFoo() {
    return this.get('foo');
  }
}
```

And you can specify default params as discussed in #1238:

```typescript
class FooAPI extends RESTDataSource {
  baseURL = 'https://api.example.com';

  defaultParams = () => ({
    api_key: this.context.token,
  });

  getFoo(a) {
    return this.get('foo', { a });
  }
}
```

So `getFoo(1)` would result in a request to `'https://api.example.com/foo?api_key=secret&a=1'`.

I also added a `defaults` callback, which allows you to specify default fetch options:
```typescript
class FooAPI extends RESTDataSource {
  baseURL = 'https://api.example.com';
  defaults = { credentials: 'include' };
}
```

One worry is that there is overlap between `defaults` and `willSendRequest`. You could use both to set headers for example:
```typescript
class FooAPI extends RESTDataSource {
  baseURL = 'https://api.example.com';
  defaults = () => ({
    headers: { 'Authorization': this.context.token }
  });
}
```

vs:

```typescript
class FooAPI extends RESTDataSource {
  baseURL = 'https://api.example.com';

  willSendRequest(request: Request) {
    request.headers.set('Authorization', this.context.token);
  }

  getFoo() {
    return this.get('foo');
  }
}
```

It seems this could get confusing.

Also, I think ideally we could get rid of `defaultParams` and rely on `willSendRequest` for this as well, but unfortunately `request.url` is a read-only string, not a `URL`. So you can't do:

```typescript
class FooAPI extends RESTDataSource {
  baseURL = 'https://api.example.com';

  willSendRequest(request: Request) {
    if (request.method === 'GET') {
      request.url.searchParams.set('api_key', this.context.token);
    } else {
      request.headers.set('Authorization', this.context.token);
    }
  }

  getFoo() {
    return this.get('foo');
  }
}
```

One option would be to pass in `RequestInit` and `URL` to `willSendRequest` instead of `Request`, but that also feels awkward.

danilobuerger commented:
I personally would get rid of `defaults` and `defaultParams` in favor of `willSendRequest` and accept the awkwardness vs having to explain to users that there are different ways to do the same thing and which one is better and so on.

martijnwalraven commented:
@danilobuerger I agree with your feedback and simplified the API. I removed `defaults` and `defaultParams`, and changed `willSendRequest` to take `RequestOptions` instead of `Request`. 

`willSendRequest` is now called before `baseURL` and has access to `path` and `params`. I also changed `RequestOptions` to include `params` as a `URLSearchParams` and `headers` as a `Headers` instance.

That means you can now do:

```typescript
  willSendRequest(request: RequestOptions) {
    if (request.method === 'GET') {
      request.params.set('api_key', this.context.token);
    } else {
      request.headers.set('Authorization', this.context.token);
    }
  }
```

What do you think?

I like allowing `baseURL` to return a `Promise` to allow for async resolving, but I'm wondering if we really need the complexity of allowing `baseURL` to be parameterized by `method`, `path` and other `RequestOptions`. Do you have a real world use case for this in mind?

martijnwalraven commented:
@danilobuerger I'm still bike shedding over this. It seems allowing `baseURL` to return a `Promise` may get a little awkward. There is no such thing as an async getter in JavaScript, so if you want to await other calls you'll have to wrap it in an immediately invoking async function. Moreover, by default it would resolve the base URL for every request, and you'd have to cache it manually to avoid this.

I was thinking a more straightforward solution might be to introduce a `resolveURL` method that takes a `RequestContext`. That method would then be called after `willSendRequest`, and has the opportunity to asynchronously resolve the URL in any way that is needed. 

The default implementation simply creates a new `URL` from `path` and `baseURL`, so you could still invoke that as part of your implementation.

I think that would turn your original example into:
```typescript
async resolveURL(request: RequestOptions) {
  if (!this.baseURL) {
    const addresses = await resolveSrv(request.path.split("/")[1] + ".service.consul");
    this.baseURL = addresses[0];
  }
  return super.resolveURL(request);
}
```

But you could also do something more complicated and circumvent the superclass implementation completely. Not sure if this makes sense, but you could store the complete list of`addresses` for example, and implement failover in that way.

Maybe I'm overcomplicating this. What do you think?

Saeris commented:
I think that change to `willSendRequest` will work just fine. I agree with you that it is a bit clunkier, but ultimately I think this makes the method more powerful overall. Particularly in that you can use it to debug exactly what will be sent at a more granular and controllable level.

I don't see an issue with `baseURL` returning a promise. If a niche use case arises, this will give people more power than what a simple string value provides. One possible scenario could be that someone might want to use a single DataSource class to handle all of their REST calls, determining a baseURL dynamically based on the endpoint being called.

martijnwalraven commented:
@Saeris The issue I saw with allowing `baseURL` to return a promise is that there isn't a way to define an async getter, and calling other async functions may get awkward.

Setting `baseURL` to a function could be an alternative, but on second thought it seemed cleaner to extract this into a separate `resolveURL` method. That also give you control over caching of base URL resolving.

So you would do:
```typescript
async resolveURL(request: RequestOptions) {
  if (!this.baseURL) {
    const addresses = await resolveSrv(request.path.split("/")[1] + ".service.consul");
    this.baseURL = addresses[0];
  }
  return super.resolveURL(request);
}
```

Instead of:
```typescript
baseURL = async (request: RequestOptions) => {
    const addresses = await resolveSrv(request.path.split("/")[1] + ".service.consul");
    return addresses[0];
}
```

What do you think?

Saeris commented:
`resolveURL` would probably be fine. I didn't have a need for a dynamic baseURL personally, so I'll defer to your judgement on that one. Perhaps the only concern is we're again heading back towards a solution that adds a new method and adds an extra step to the process which may confuse users.

danilobuerger commented:
A bit late now, but why is `willSendRequest` even necessary now with having `resolveURL` ?

martijnwalraven commented:
@danilobuerger Although technically you could do everything in `resolveURL`, I think it still makes sense to have separate callbacks for clarity.

I see `resolveURL` as something most people won't need to override, except if they need dynamic async resolving.

Implementing `willSendRequest` is pretty common on the other hand, because that's where you add headers or parameters. And I want to make that as easy as possible (so no need to call super for instance).

danilobuerger commented:
Ok, sure I can see that.

martijnwalraven commented:
@danilobuerger An alternative might be to allow `willSendRequest` to return a `URL` (or a complete `Request`). Not sure if that makes things clearer though.

danilobuerger commented:
Hmm, all speculation at that point. Maybe lets just leave it as is now and see how other people feel about it when they start using it?

martijnwalraven commented:
Yeah, you're right, I know I'm prone to bike shedding :)
